File: Python/Priority queue/Take Gifts From the Richest Pile.py
class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        pq=[]
        for i in gifts:
            heapq.heappush(pq, -i)
        
        while k!=0:
            k -=1
            t = -heapq.heappop(pq)
            heapq.heappush(pq,-math.floor(math.sqrt(t)))
        sum =0
        while pq:
            sum+=-heapq.heappop(pq)
        return sum

File: CPP/Priority queue/Take Gifts From the Richest Pile.cpp
class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k) {
        priority_queue<int>pq;
        for(auto i:gifts){
            pq.push(i);
        }
        while(k!=0){
            int t = pq.top();
            pq.pop();
            pq.push(floor(sqrt(t)));
            k--;
        }
        long long sum=0;
        while(!pq.empty()){
            sum+=pq.top();pq.pop();
        }
        return sum;
    }
};
